File: src/java/com/xtaudio/xt/XtDevice.java
package com.xtaudio.xt;

import com.sun.jna.Native;
import com.sun.jna.Pointer;
import com.xtaudio.xt.CoreStructs.DeviceStreamParams;
import com.xtaudio.xt.CoreStructs.StreamParams;
import com.xtaudio.xt.CoreStructs.XtBufferSize;
import com.xtaudio.xt.CoreStructs.XtDeviceStreamParams;
import com.xtaudio.xt.CoreStructs.XtFormat;
import com.xtaudio.xt.CoreStructs.XtMix;
import java.nio.charset.Charset;
import java.util.Optional;

import com.sun.jna.ptr.IntByReference;
import com.sun.jna.ptr.PointerByReference;
import static com.xtaudio.xt.Utility.handleError;

public final class XtDevice implements AutoCloseable {

    static { Native.register(Utility.LIBRARY); }
    private static native void XtDeviceDestroy(Pointer d);
    private static native long XtDeviceShowControlPanel(Pointer d);
    private static native long XtDeviceGetMix(Pointer d, IntByReference valid, XtMix mix);
    private static native long XtDeviceGetName(Pointer d, byte[] buffer, IntByReference size);
    private static native long XtDeviceGetBufferSize(Pointer d, XtFormat format, XtBufferSize size);
    private static native long XtDeviceGetChannelCount(Pointer d, boolean output, IntByReference count);
    private static native long XtDeviceSupportsFormat(Pointer d, XtFormat format, IntByReference supports);
    private static native long XtDeviceSupportsAccess(Pointer d, boolean interleaved, IntByReference supports);
    private static native long XtDeviceGetChannelName(Pointer d, boolean output, int index, byte[] buffer, IntByReference size);
    private static native long XtDeviceOpenStream(Pointer d, DeviceStreamParams params, Pointer user, PointerByReference stream);

    private final Pointer _d;
    Pointer handle() { return _d; }
    XtDevice(Pointer d) { _d = d; }

    @Override public void close() { XtDeviceDestroy(_d); }
    @Override public String toString() { return getName(); }
    public void showControlPanel() { handleError(XtDeviceShowControlPanel(_d)); }

    public XtBufferSize getBufferSize(XtFormat format) {
        var result = new XtBufferSize();
        handleError(XtDeviceGetBufferSize(_d, format, result));
        return result;
    }

    public int getChannelCount(boolean output) {
        var count = new IntByReference();
        handleError(XtDeviceGetChannelCount(_d, output, count));
        return count.getValue();
    }

    public boolean supportsFormat(XtFormat format) {
        var supports = new IntByReference();
        handleError(XtDeviceSupportsFormat(_d, format, supports));
        return supports.getValue() != 0;
    }

    public boolean supportsAccess(boolean interleaved) {
        var supports = new IntByReference();
        handleError(XtDeviceSupportsAccess(_d, interleaved, supports));
        return supports.getValue() != 0;
    }

    public Optional<XtMix> getMix() {
        XtMix mix = new XtMix();
        var valid = new IntByReference();
        handleError(XtDeviceGetMix(_d, valid, mix));
        return valid.getValue() == 0? Optional.empty(): Optional.of(mix);
    }

    public String getName() {
        var size = new IntByReference();
        handleError(XtDeviceGetName(_d, null, size));
        byte[] buffer = new byte[size.getValue()];
        handleError(XtDeviceGetName(_d, buffer, size));
        return new String(buffer, 0, size.getValue() - 1, Charset.forName("UTF-8"));
    }

    public String getChannelName(boolean output, int index) {
        var size = new IntByReference();
        handleError(XtDeviceGetChannelName(_d, output, index, null, size));
        byte[] buffer = new byte[size.getValue()];
        handleError(XtDeviceGetChannelName(_d, output, index, buffer, size));
        return new String(buffer, 0, size.getValue() - 1, Charset.forName("UTF-8"));
    }

    public XtStream openStream(XtDeviceStreamParams params, Object user) {
        var stream = new PointerByReference();
        var result = new XtStream(params.stream.onBuffer, params.stream.onXRun, user);
        var native_ = new DeviceStreamParams();
        native_.format = params.format;
        native_.stream = new StreamParams();
        native_.bufferSize = params.bufferSize;
        native_.stream.onBuffer = result.onNativeBuffer();
        native_.stream.interleaved = params.stream.interleaved;
        native_.stream.onXRun = params.stream.onXRun == null? null: result.onNativeXRun();
        handleError(XtDeviceOpenStream(_d, native_, Pointer.NULL, stream));
        result.init(stream.getValue());
        return result;
    }
}

File: src/core/xt/xt/audio/XtDevice.cpp
#include <xt/audio/XtDevice.h>
#include <xt/Private.hpp>
#include <cstring>

void XT_CALL
XtDeviceDestroy(XtDevice* d) 
{
  XT_ASSERT(XtiCalledOnMainThread());
  delete d;
}

XtError XT_CALL
XtDeviceShowControlPanel(XtDevice* d) 
{
  XT_ASSERT(d != nullptr);
  XT_ASSERT(XtiCalledOnMainThread());
  return XtiCreateError(d->GetSystem(), d->ShowControlPanel());
}

XtError XT_CALL
XtDeviceGetMix(XtDevice const* d, XtBool* valid, XtMix* mix)
{
  XT_ASSERT(d != nullptr);
  XT_ASSERT(mix != nullptr);
  XT_ASSERT(XtiCalledOnMainThread());
  *valid = XtFalse;
  std::memset(mix, 0, sizeof(XtMix));
  return XtiCreateError(d->GetSystem(), d->GetMix(valid, mix));
}

XtError XT_CALL
XtDeviceGetName(XtDevice const* d, char* buffer, int32_t* size)
{
  XT_ASSERT(d != nullptr);
  XT_ASSERT(XtiCalledOnMainThread());
  XT_ASSERT(size != nullptr && *size >= 0);
  return XtiCreateError(d->GetSystem(), d->GetName(buffer, size));
}

XtError XT_CALL
XtDeviceGetChannelCount(XtDevice const* d, XtBool output, int32_t* count)
{
  XT_ASSERT(d != nullptr);
  XT_ASSERT(count != nullptr);
  XT_ASSERT(XtiCalledOnMainThread());
  *count = 0;
  return XtiCreateError(d->GetSystem(), d->GetChannelCount(output, count));
}

XtError XT_CALL
XtDeviceSupportsAccess(XtDevice const* d, XtBool interleaved, XtBool* supports)
{
  XT_ASSERT(d != nullptr);
  XT_ASSERT(supports != nullptr);
  XT_ASSERT(XtiCalledOnMainThread());
  *supports = XtFalse;
  return XtiCreateError(d->GetSystem(), d->SupportsAccess(interleaved, supports));
}

XtError XT_CALL
XtDeviceSupportsFormat(XtDevice const* d, XtFormat const* format, XtBool* supports)
{
  XT_ASSERT(d != nullptr);
  XT_ASSERT(format != nullptr);
  XT_ASSERT(supports != nullptr);
  XT_ASSERT(XtiCalledOnMainThread());
  XT_ASSERT(XtiValidateFormat(d->GetSystem(), *format));
  *supports = XtFalse;
  return XtiCreateError(d->GetSystem(), d->SupportsFormat(format, supports));
}

XtError XT_CALL
XtDeviceGetBufferSize(XtDevice const* d, XtFormat const* format, XtBufferSize* size)
{
  XT_ASSERT(d != nullptr);
  XT_ASSERT(size != nullptr);
  XT_ASSERT(format != nullptr);
  XT_ASSERT(XtiCalledOnMainThread());
  XT_ASSERT(XtiValidateFormat(d->GetSystem(), *format));
  XtError error;
  XtBool supports;
  XtSystem system = d->GetSystem();
  std::memset(size, 0, sizeof(XtBufferSize));
  if((error = XtDeviceSupportsFormat(d, format, &supports)) != 0) return error;
  if(!supports) return XtiCreateError(system, XtAudioGetService(system)->GetFormatFault());
  return XtiCreateError(d->GetSystem(), d->GetBufferSize(format, size));
}

XtError XT_CALL
XtDeviceGetChannelName(XtDevice const* d, XtBool output, int32_t index, char* buffer, int32_t* size)
{
  XT_ASSERT(index >= 0);
  XT_ASSERT(d != nullptr);
  XT_ASSERT(XtiCalledOnMainThread());  
  XT_ASSERT(size != nullptr && *size >= 0);
  return XtiCreateError(d->GetSystem(), d->GetChannelName(output, index, buffer, size));
}